onnx/shape_inference/implementation.cc

```diff
@@ -115,6 +115,39 @@ void checkShapesAndTypes(const TypeProto& inferredType, const TypeProto& existin
     fail_type_inference("type case unsupported. existing=", existingTypeCase, " inferred=", inferredTypeCase);
   }
 }
+// TypeProto_Tensor or TypeProto_SparseTensor
+template <typename T>
+void generateSymbolicShape(T inferredType, T* symbolicInferredType, SymbolicShape symbolicShape) {
+  if (!inferredType.has_shape()) {
+    *symbolicInferredType = inferredType;
+    return;
+  }
+  for (int i = 0; i < inferredType.shape().dim_size(); ++i) {
+    const auto& inferredDim = inferredType.shape().dim(i);
+    auto* typeAfterSymbolicShapeDim = symbolicInferredType->mutable_shape()->mutable_dim(i);
+    *typeAfterSymbolicShapeDim = inferredDim;
+    // If shape cannot be inferred; set it as a symbolic shape
+    if (inferredType.shape().dim(i).dim_value() == 0 && !inferredType.shape().dim(i).has_dim_param()) {
```

Comment:
Shouldn't first condition be "! has_dim_value()"?

Reply:
You are right. Besides this one, shall I add another condition: `inferredType.shape().dim(i).has_dim_param() && inferredType.shape().dim(i).dim_value() == -1` ?